### DISCUSSION TITLE
Changing nullability from `T?` to `Option<T>`

### DISCUSSION THREAD

**Discussion by milgner** [ExpediaGroup/graphql-kotlin#1823] [Q&A]
In the process of adopting [Arrow](https://arrow-kt.io/) and representing nullable fields using `Option`, I realised that my data classes don't work well with GraphQL anymore.

I wonder whether it's possible to change the way that `null` values are treated? Ideally I'd like to register a type conversion so that a field of type `Option<T>` works the same way as `T?`.

**Comment by milgner** [comment]
After some cursory browsing, it seems like I'd need to change [`convertArgumentValue`](https://github.com/ExpediaGroup/graphql-kotlin/blob/405255427937306dc1e6326d92b236966476e79b/generator/graphql-kotlin-schema-generator/src/main/kotlin/com/expediagroup/graphql/generator/execution/convertArgumentValue.kt#L85) and check for `paramType.getKClass() == Option::class` or something along that line? But since this function is internal, it looks like there's no access to it?

**Reply by dariuszkuc** [reply]
Yes, in order to support custom wrapper objects (i.e. `Option`) you would need to update both schema generator hooks so you can generate valid schema and you would also need a custom data fetcher that knows how to handle incoming variables. 

Sadly, as you noticed, currently there is no easy way to extend default data fetcher and you would need to basically duplicate the code locally with additional logic to handle it. Corresponding issue https://github.com/ExpediaGroup/graphql-kotlin/issues/1445
